src/Cli/dotnet/Commands/Package/Add/PackageAddCommand.cs

```diff
@@ -133,4 +139,214 @@ private string[] TransformArgs(PackageIdentityWithRange packageId, string tempDg
+                editor.SourceFile.Save();
+            }
+        }
+
+        (Action Revert, Action<string> Update, Action Save)? SetCentralVersion(string version)
+        {
+            // Find out whether CPM is enabled.
+            if (!string.Equals(projectInstance.GetProperty("ManagePackageVersionsCentrally")?.EvaluatedValue, bool.TrueString, StringComparison.OrdinalIgnoreCase))
+            {
+                return null;
+            }
+
+            // Load the Directory.Packages.props project.
+            var directoryPackagesPropsPath = projectInstance.GetProperty("DirectoryPackagesPropsPath")?.EvaluatedValue;
+            if (!File.Exists(directoryPackagesPropsPath))
+            {
+                Reporter.Verbose.WriteLine($"Directory.Packages.props file does not exist: {directoryPackagesPropsPath}");
+                return null;
+            }
+
+            var snapshot = File.ReadAllText(directoryPackagesPropsPath);
+            var directoryPackagesPropsProject = projectCollection.LoadProject(directoryPackagesPropsPath);
+
+            const string packageVersionItemType = "PackageVersion";
+            const string versionAttributeName = "Version";
+
+            // Update existing PackageVersion if it exists.
+            var packageVersion = directoryPackagesPropsProject.GetItems(packageVersionItemType)
+                .LastOrDefault(i => string.Equals(i.EvaluatedInclude, _packageId.Id, StringComparison.OrdinalIgnoreCase));
+            if (packageVersion != null)
+            {
+                var packageVersionItemElement = packageVersion.Project.GetItemProvenance(packageVersion).LastOrDefault()?.ItemElement;
+                var versionAttribute = packageVersionItemElement?.Metadata.FirstOrDefault(i => i.Name.Equals(versionAttributeName, StringComparison.OrdinalIgnoreCase));
+                if (versionAttribute != null)
+                {
+                    versionAttribute.Value = version;
+                    directoryPackagesPropsProject.Save();
+
+                    // If user didn't specify a version and a version is already specified in Directory.Packages.props,
+                    // don't update the Directory.Packages.props (that's how the project-based equivalent behaves as well).
+                    if (!hasVersion)
+                    {
+                        skipUpdate = true;
+                        return (Revert: NoOp, Update: Unreachable, Save: Revert);
+
+                        static void NoOp() { }
+                        static void Unreachable(string value) => Debug.Fail("Unreachable.");
+                    }
+
+                    return (Revert, Update, Save);
+
+                    void Update(string value)
+                    {
+                        versionAttribute.Value = value;
+                        directoryPackagesPropsProject.Save();
+                    }
+                }
+            }
+
+            {
```

Comment:
I know you can make arbitrary scopes like this, but what is this for?

Reply:
This allows me to define the `void Update(string value)` local function twice - otherwise the nested one would "see" the parent one and complain it's re-declared.

Reply:
I would say that this is rather confusing: consider either using different names, such as `UpdateInVersion`/`UpdateInMetadata`, or combining the local function into one that takes an extra parameter, and then wrapping them in a lambda at the return to pass either `versionAttribute` or `metadata`.